### DISCUSSION TITLE
Dot notation for types

### DISCUSSION THREAD

**Discussion by gustavopch** [DanielXMoore/Civet#190] [Ideas]
```ts
type Bar = Foo.bar
type Bar = Foo!.bar
type Bar = Foo?.bar
type Bar = Foo?.bar ?? number
↓↓↓
type Bar = Foo['bar']
type Bar = NonNullable<Foo>['bar']
type Bar = NonNullable<Foo>['bar'] | undefined
type Bar = NonNullable<Foo>['bar'] | number
```

**Comment by STRd6** [comment]
I like it!

**Comment by orenelbaum** [comment]
I think that it breaks namespaces?

**Reply by gustavopch** [reply]
@STRd6 Is it possible to distinguish between types/interfaces and namespaces?

**Reply by STRd6** [reply]
https://www.typescriptlang.org/docs/handbook/namespaces-and-modules.html#using-modules
> It is also worth noting that, for Node.js applications, modules are the default and we recommended modules over namespaces in modern code.

We can add a `"civet tsCompat"` or `"civet tsNamespaces"` to add compatibility for legacy TS code using namespaces. Fortunately we don't have the same constraints as TS does about backward compatibility.

**Reply by orenelbaum** [reply]
I think that namespaces are still very widely used in libraries. Both React and Solid use `JSX.Element`.

**Reply by edemaine** [reply]
Ideally we could detect that `Foo` is a namespace, but I'm afraid that requires compile-time type understanding. 😢

**Reply by orenelbaum** [reply]
Yeah I think that it's probably not practical

**Reply by STRd6** [reply]
We could add configuration based namespace whitelists for each popular library... but that would be a ton of work. I do think this is a good idea for the longer term and eventually there are things we'd like to do that require more advanced compile-time type/file processing. It's probably worth exploring (at least as an experimental option) to see what the practical complications are.

**Reply by orenelbaum** [reply]
Personally I don't use type indexing a lot so it seems to me like it's probably not worth the complexity. But maybe you guys are using type indexing more than me.

**Reply by STRd6** [reply]
If this does end up being generally impractical we can see about extending type destructuring to cover the same use cases.
